### PR TITLE
GEODE-4604: rest api function execution needs to check the required p…

### PR DESCRIPTION
**Pull request by jinmeiliao**
…ermission of each function

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java [hunk 1]
```diff
@@ -62,61 +58,58 @@ public static void before() throws Exception {
...
     assertEquals(401, GeodeRestClient.getCode(response));
-    response = restClient.doPost("/functions/AddFreeItemsToOrder", "dataReader", "1234567", json);
-    assertEquals(403, GeodeRestClient.getCode(response));
-    response = restClient.doPost("/functions/AddFreeItemsToOrder?onRegion=" + REGION_NAME,
-        "dataWriter", "1234567", json);
-    // because we're only testing the security of the endpoint, not the endpoint functionality, a
-    // 500 is acceptable
-    assertEquals(500, GeodeRestClient.getCode(response));
+    response = restClient.doPost("/functions/invalid-function-id", "user", "user", "");
+    assertEquals(404, GeodeRestClient.getCode(response));
```
Comment (pdxrunner) [inline geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java]: Shouldn't this be a "403"? Test failed when I tried running it.

File: geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java [hunk 2]
```diff
@@ -62,61 +58,58 @@ public static void before() throws Exception {
...
+    response = restClient.doGet("/functions", "dataRead", "dataRead");
     assertEquals(200, GeodeRestClient.getCode(response));
     response.getEntity();
     assertThat(GeodeRestClient.getContentType(response))
         .containsIgnoringCase(MediaType.APPLICATION_JSON_UTF8_VALUE);
+  }
 
-    response = restClient.doPost("/functions/AddFreeItemsToOrder", "unknown-user", "1234567", json);
+  @Test
+  public void executeFunction() throws Exception {
```
Comment (pdxrunner) [inline geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java]: A better name for this test would be `executeFunctionNotAuthorized` or something similar to note the the test is only running functions executions that will fail.

File: geode-web-api/src/main/java/org/apache/geode/rest/internal/web/security/RestSecurityService.java [hunk 3]
```diff
@@ -46,6 +47,11 @@ public boolean authorize(String resource, String operation, String region) {
     return authorize(resource, operation, region, null);
   }
 
+  public void authorize(ResourcePermission permission) {
```
Comment (jdeppe-pivotal) [inline geode-web-api/src/main/java/org/apache/geode/rest/internal/web/security/RestSecurityService.java]: Shouldn't this return a boolean?

Comment (jinmeiliao) [reply geode-web-api/src/main/java/org/apache/geode/rest/internal/web/security/RestSecurityService.java]: a boolean return value is required only if we are using the method in the PreAuthorize tag on the controller. e.g.
 @PreAuthorize("@securityService.authorize('DATA', 'READ')")

We are only using this method inside the controller body.

File: geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java (hunk 1 above)
Comment (jinmeiliao) [reply geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java]: you will need to rebuild the war file before running the test.

Comment (pdxrunner) [reply geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityIntegrationTest.java]: Thanks, my oversight.

**Comment by pdxrunner** [comment]
Looks OK other than spotless

**Comment by jinmeiliao** [comment]
precheckin green

**Pull request merged**